File: fixtures/type-limits/tests/bindings/test_type_limits.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from uniffi_type_limits import *

import math
import unittest

class TestTypeLimits(unittest.TestCase):
    def test_strict_lower_bounds(self):
        self.assertRaises(ValueError, lambda: take_i8(-2**7 - 1))
        self.assertRaises(ValueError, lambda: take_i16(-2**15 - 1))
        self.assertRaises(ValueError, lambda: take_i32(-2**31 - 1))
        self.assertRaises(ValueError, lambda: take_i64(-2**63 - 1))
        self.assertRaises(ValueError, lambda: take_u8(-1))
        self.assertRaises(ValueError, lambda: take_u16(-1))
        self.assertRaises(ValueError, lambda: take_u32(-1))
        self.assertRaises(ValueError, lambda: take_u64(-1))

        self.assertEqual(take_i8(-2**7), -2**7)
        self.assertEqual(take_i16(-2**15), -2**15)
        self.assertEqual(take_i32(-2**31), -2**31)
        self.assertEqual(take_i64(-2**63), -2**63)
        self.assertEqual(take_u8(0), 0)
        self.assertEqual(take_u16(0), 0)
        self.assertEqual(take_u32(0), 0)
        self.assertEqual(take_u64(0), 0)

    def test_strict_upper_bounds(self):
        self.assertRaises(ValueError, lambda: take_i8(2**7))
        self.assertRaises(ValueError, lambda: take_i16(2**15))
        self.assertRaises(ValueError, lambda: take_i32(2**31))
        self.assertRaises(ValueError, lambda: take_i64(2**63))
        self.assertRaises(ValueError, lambda: take_u8(2**8))
        self.assertRaises(ValueError, lambda: take_u16(2**16))
        self.assertRaises(ValueError, lambda: take_u32(2**32))
        self.assertRaises(ValueError, lambda: take_u64(2**64))

        self.assertEqual(take_i8(2**7 - 1), 2**7 - 1)
        self.assertEqual(take_i16(2**15 - 1), 2**15 - 1)
        self.assertEqual(take_i32(2**31 - 1), 2**31 - 1)
        self.assertEqual(take_i64(2**63 - 1), 2**63 - 1)
        self.assertEqual(take_u8(2**8 - 1), 2**8 - 1)
        self.assertEqual(take_u16(2**16 - 1), 2**16 - 1)
        self.assertEqual(take_u32(2**32 - 1), 2**32 - 1)
        self.assertEqual(take_u64(2**64 - 1), 2**64 - 1)

    def test_larger_numbers(self):
        self.assertRaises(ValueError, lambda: take_i8(10**3))
        self.assertRaises(ValueError, lambda: take_i16(10**5))
        self.assertRaises(ValueError, lambda: take_i32(10**10))
        self.assertRaises(ValueError, lambda: take_i64(10**19))
        self.assertRaises(ValueError, lambda: take_u8(10**3))
        self.assertRaises(ValueError, lambda: take_u16(10**5))
        self.assertRaises(ValueError, lambda: take_u32(10**10))
        self.assertRaises(ValueError, lambda: take_u64(10**20))

        self.assertEqual(take_i8(10**2), 10**2)
        self.assertEqual(take_i16(10**4), 10**4)
        self.assertEqual(take_i32(10**9), 10**9)
        self.assertEqual(take_i64(10**18), 10**18)
        self.assertEqual(take_u8(10**2), 10**2)
        self.assertEqual(take_u16(10**4), 10**4)
        self.assertEqual(take_u32(10**9), 10**9)
        self.assertEqual(take_u64(10**19), 10**19)

    def test_non_integer(self):
        self.assertRaises(TypeError, lambda: take_i8(None))
        self.assertRaises(TypeError, lambda: take_i16(None))
        self.assertRaises(TypeError, lambda: take_i32(None))
        self.assertRaises(TypeError, lambda: take_i64(None))
        self.assertRaises(TypeError, lambda: take_u8(None))
        self.assertRaises(TypeError, lambda: take_u16(None))
        self.assertRaises(TypeError, lambda: take_u32(None))
        self.assertRaises(TypeError, lambda: take_u64(None))

        self.assertRaises(TypeError, lambda: take_i8("0"))
        self.assertRaises(TypeError, lambda: take_i16("0"))
        self.assertRaises(TypeError, lambda: take_i32("0"))
        self.assertRaises(TypeError, lambda: take_i64("0"))
        self.assertRaises(TypeError, lambda: take_u8("0"))
        self.assertRaises(TypeError, lambda: take_u16("0"))
        self.assertRaises(TypeError, lambda: take_u32("0"))
        self.assertRaises(TypeError, lambda: take_u64("0"))

        self.assertRaises(TypeError, lambda: take_i8(0.0))
        self.assertRaises(TypeError, lambda: take_i16(0.0))
        self.assertRaises(TypeError, lambda: take_i32(0.0))
        self.assertRaises(TypeError, lambda: take_i64(0.0))
        self.assertRaises(TypeError, lambda: take_u8(0.0))
        self.assertRaises(TypeError, lambda: take_u16(0.0))
        self.assertRaises(TypeError, lambda: take_u32(0.0))
        self.assertRaises(TypeError, lambda: take_u64(0.0))

        class A:
            pass

        self.assertRaises(TypeError, lambda: take_i8(A()))
        self.assertRaises(TypeError, lambda: take_i16(A()))
        self.assertRaises(TypeError, lambda: take_i32(A()))
        self.assertRaises(TypeError, lambda: take_i64(A()))
        self.assertRaises(TypeError, lambda: take_u8(A()))
        self.assertRaises(TypeError, lambda: take_u16(A()))
        self.assertRaises(TypeError, lambda: take_u32(A()))
        self.assertRaises(TypeError, lambda: take_u64(A()))

    def test_integer_like(self):
        self.assertEqual(take_i8(123), 123.0)
        self.assertEqual(take_i16(123), 123.0)
        self.assertEqual(take_i32(123), 123.0)
        self.assertEqual(take_i64(123), 123.0)
        self.assertEqual(take_u8(123), 123.0)
        self.assertEqual(take_u16(123), 123.0)
        self.assertEqual(take_u32(123), 123.0)
        self.assertEqual(take_u64(123), 123.0)

        self.assertEqual(take_i8(False), 0)
        self.assertEqual(take_i16(False), 0)
        self.assertEqual(take_i32(False), 0)
        self.assertEqual(take_i64(False), 0)
        self.assertEqual(take_u8(False), 0)
        self.assertEqual(take_u16(False), 0)
        self.assertEqual(take_u32(False), 0)
        self.assertEqual(take_u64(False), 0)

        self.assertEqual(take_i8(True), 1)
        self.assertEqual(take_i16(True), 1)
        self.assertEqual(take_i32(True), 1)
        self.assertEqual(take_i64(True), 1)
        self.assertEqual(take_u8(True), 1)
        self.assertEqual(take_u16(True), 1)
        self.assertEqual(take_u32(True), 1)
        self.assertEqual(take_u64(True), 1)

        class A:
            def __index__(self):
                return 123

        self.assertEqual(take_i8(A()), 123)
        self.assertEqual(take_i16(A()), 123)
        self.assertEqual(take_i32(A()), 123)
        self.assertEqual(take_i64(A()), 123)
        self.assertEqual(take_u8(A()), 123)
        self.assertEqual(take_u16(A()), 123)
        self.assertEqual(take_u32(A()), 123)
        self.assertEqual(take_u64(A()), 123)

    def test_non_float(self):
        self.assertRaises(TypeError, lambda: take_f32(None))
        self.assertRaises(TypeError, lambda: take_f64(None))

        self.assertRaises(TypeError, lambda: take_f32("0"))
        self.assertRaises(TypeError, lambda: take_f64("0"))

        self.assertRaises(TypeError, lambda: take_f32(1j))
        self.assertRaises(TypeError, lambda: take_f64(1j))

        class A:
            pass

        self.assertRaises(TypeError, lambda: take_f32(A()))
        self.assertRaises(TypeError, lambda: take_f64(A()))

    def test_float_like(self):
        self.assertEqual(take_f32(False), 0.0)
        self.assertEqual(take_f64(False), 0.0)

        self.assertEqual(take_f32(True), 1.0)
        self.assertEqual(take_f64(True), 1.0)

        self.assertEqual(take_f32(123), 123.0)
        self.assertEqual(take_f64(123), 123.0)

        class A:
            def __float__(self):
                return 456.0

        self.assertEqual(take_f32(A()), 456.0)
        self.assertEqual(take_f64(A()), 456.0)

    def test_special_floats(self):
        self.assertEqual(take_f32(math.inf), math.inf)
        self.assertEqual(take_f64(math.inf), math.inf)

        self.assertEqual(take_f32(-math.inf), -math.inf)
        self.assertEqual(take_f64(-math.inf), -math.inf)

        self.assertEqual(math.copysign(1.0, take_f32(0.0)), 1.0)
        self.assertEqual(math.copysign(1.0, take_f64(0.0)), 1.0)

        self.assertEqual(math.copysign(1.0, take_f32(-0.0)), -1.0)
        self.assertEqual(math.copysign(1.0, take_f64(-0.0)), -1.0)

        self.assertTrue(math.isnan(take_f32(math.nan)))
        self.assertTrue(math.isnan(take_f64(math.nan)))

    def test_non_string(self):
        self.assertRaises(TypeError, lambda: take_string(None))
        self.assertRaises(TypeError, lambda: take_string(False))
        self.assertRaises(TypeError, lambda: take_string(True))
        self.assertRaises(TypeError, lambda: take_string(0))
        self.assertRaises(TypeError, lambda: take_string(0.0))
        self.assertRaises(TypeError, lambda: take_string(b""))

        class A:
            def __str__(self):
                return ""

        self.assertRaises(TypeError, lambda: take_string(A()))

    def test_strings(self):
        self.assertRaises(ValueError, lambda: take_string("\ud800")) # surrogate
        self.assertEqual(take_string(""), "")
        self.assertEqual(take_string("愛"), "愛")
        self.assertEqual(take_string("💖"), "💖")

    def test_non_bytes(self):
        self.assertRaises(TypeError, lambda: take_bytes(None))
        self.assertRaises(TypeError, lambda: take_bytes(False))
        self.assertRaises(TypeError, lambda: take_bytes(True))
        self.assertRaises(TypeError, lambda: take_bytes(0))
        self.assertRaises(TypeError, lambda: take_bytes(0.0))
        self.assertRaises(TypeError, lambda: take_bytes(""))

        class A:
            def __str__(self):
                return ""

        self.assertRaises(TypeError, lambda: take_bytes(A()))

    def test_bytes(self):
        self.assertEqual(take_bytes(b""), b"")
        self.assertEqual(take_bytes(b"\xff"), b"\xff") # invalid utf-8 byte
        self.assertEqual(take_bytes(b"\xed\xa0\x80"), b"\xed\xa0\x80") # surrogate
        self.assertEqual(take_bytes("愛".encode()), "愛".encode())
        self.assertEqual(take_bytes("💖".encode()), "💖".encode())
        self.assertEqual(take_bytes("愛".encode("utf-16-le")), b"\x1b\x61")
        self.assertEqual(take_bytes("💖".encode("utf-16-le")), b"\x3d\xd8\x96\xdc")

if __name__ == "__main__":
    unittest.main()

File: fixtures/type-limits/tests/bindings/test_type_limits.rb
# frozen_string_literal: true

# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/. */

require 'test/unit'
require 'uniffi_type_limits'

class TestTypeLimits < Test::Unit::TestCase
  def test_strict_lower_bounds
    assert_raise RangeError do UniffiTypeLimits.take_i8(-2**7 - 1) end
    assert_raise RangeError do UniffiTypeLimits.take_i16(-2**15 - 1) end
    assert_raise RangeError do UniffiTypeLimits.take_i32(-2**31 - 1) end
    assert_raise RangeError do UniffiTypeLimits.take_i64(-2**63 - 1) end
    assert_raise RangeError do UniffiTypeLimits.take_u8(-1) end
    assert_raise RangeError do UniffiTypeLimits.take_u16(-1) end
    assert_raise RangeError do UniffiTypeLimits.take_u32(-1) end
    assert_raise RangeError do UniffiTypeLimits.take_u64(-1) end

    assert_equal(UniffiTypeLimits.take_i8(-2**7), -2**7)
    assert_equal(UniffiTypeLimits.take_i16(-2**15), -2**15)
    assert_equal(UniffiTypeLimits.take_i32(-2**31), -2**31)
    assert_equal(UniffiTypeLimits.take_i64(-2**63), -2**63)
    assert_equal(UniffiTypeLimits.take_u8(0), 0)
    assert_equal(UniffiTypeLimits.take_u16(0), 0)
    assert_equal(UniffiTypeLimits.take_u32(0), 0)
    assert_equal(UniffiTypeLimits.take_u64(0), 0)
  end
  def test_strict_upper_bounds
    assert_raise RangeError do UniffiTypeLimits.take_i8(2**7) end
    assert_raise RangeError do UniffiTypeLimits.take_i16(2**15) end
    assert_raise RangeError do UniffiTypeLimits.take_i32(2**31) end
    assert_raise RangeError do UniffiTypeLimits.take_i64(2**63) end
    assert_raise RangeError do UniffiTypeLimits.take_u8(2**8) end
    assert_raise RangeError do UniffiTypeLimits.take_u16(2**16) end
    assert_raise RangeError do UniffiTypeLimits.take_u32(2**32) end
    assert_raise RangeError do UniffiTypeLimits.take_u64(2**64) end

    assert_equal(UniffiTypeLimits.take_i8(2**7 - 1), 2**7 - 1)
    assert_equal(UniffiTypeLimits.take_i16(2**15 - 1), 2**15 - 1)
    assert_equal(UniffiTypeLimits.take_i32(2**31 - 1), 2**31 - 1)
    assert_equal(UniffiTypeLimits.take_i64(2**63 - 1), 2**63 - 1)
    assert_equal(UniffiTypeLimits.take_u8(2**8 - 1), 2**8 - 1)
    assert_equal(UniffiTypeLimits.take_u16(2**16 - 1), 2**16 - 1)
    assert_equal(UniffiTypeLimits.take_u32(2**32 - 1), 2**32 - 1)
    assert_equal(UniffiTypeLimits.take_u64(2**64 - 1), 2**64 - 1)
  end
  def test_larger_numbers
    assert_raise RangeError do UniffiTypeLimits.take_i8(10**3) end
    assert_raise RangeError do UniffiTypeLimits.take_i16(10**5) end
    assert_raise RangeError do UniffiTypeLimits.take_i32(10**10) end
    assert_raise RangeError do UniffiTypeLimits.take_i64(10**19) end
    assert_raise RangeError do UniffiTypeLimits.take_u8(10**3) end
    assert_raise RangeError do UniffiTypeLimits.take_u16(10**5) end
    assert_raise RangeError do UniffiTypeLimits.take_u32(10**10) end
    assert_raise RangeError do UniffiTypeLimits.take_u64(10**20) end

    assert_equal(UniffiTypeLimits.take_i8(10**2), 10**2)
    assert_equal(UniffiTypeLimits.take_i16(10**4), 10**4)
    assert_equal(UniffiTypeLimits.take_i32(10**9), 10**9)
    assert_equal(UniffiTypeLimits.take_i64(10**18), 10**18)
    assert_equal(UniffiTypeLimits.take_u8(10**2), 10**2)
    assert_equal(UniffiTypeLimits.take_u16(10**4), 10**4)
    assert_equal(UniffiTypeLimits.take_u32(10**9), 10**9)
    assert_equal(UniffiTypeLimits.take_u64(10**19), 10**19)

    assert_raise FloatDomainError do UniffiTypeLimits.take_i8(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i16(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i32(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i64(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u8(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u16(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u32(-Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u64(-Float::INFINITY) end

    assert_raise FloatDomainError do UniffiTypeLimits.take_i8(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i16(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i32(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i64(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u8(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u16(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u32(Float::INFINITY) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u64(Float::INFINITY) end

    assert_raise FloatDomainError do UniffiTypeLimits.take_i8(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i16(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i32(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_i64(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u8(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u16(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u32(Float::NAN) end
    assert_raise FloatDomainError do UniffiTypeLimits.take_u64(Float::NAN) end
  end
  class NonInteger
  end
  def test_non_integer
    assert_raise TypeError do UniffiTypeLimits.take_i8(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_i16(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_i32(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_i64(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_u8(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_u16(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_u32(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_u64(nil) end

    assert_raise TypeError do UniffiTypeLimits.take_i8("0") end
    assert_raise TypeError do UniffiTypeLimits.take_i16("0") end
    assert_raise TypeError do UniffiTypeLimits.take_i32("0") end
    assert_raise TypeError do UniffiTypeLimits.take_i64("0") end
    assert_raise TypeError do UniffiTypeLimits.take_u8("0") end
    assert_raise TypeError do UniffiTypeLimits.take_u16("0") end
    assert_raise TypeError do UniffiTypeLimits.take_u32("0") end
    assert_raise TypeError do UniffiTypeLimits.take_u64("0") end

    assert_raise TypeError do UniffiTypeLimits.take_i8(false) end
    assert_raise TypeError do UniffiTypeLimits.take_i16(false) end
    assert_raise TypeError do UniffiTypeLimits.take_i32(false) end
    assert_raise TypeError do UniffiTypeLimits.take_i64(false) end
    assert_raise TypeError do UniffiTypeLimits.take_u8(false) end
    assert_raise TypeError do UniffiTypeLimits.take_u16(false) end
    assert_raise TypeError do UniffiTypeLimits.take_u32(false) end
    assert_raise TypeError do UniffiTypeLimits.take_u64(false) end

    assert_raise TypeError do UniffiTypeLimits.take_i8(true) end
    assert_raise TypeError do UniffiTypeLimits.take_i16(true) end
    assert_raise TypeError do UniffiTypeLimits.take_i32(true) end
    assert_raise TypeError do UniffiTypeLimits.take_i64(true) end
    assert_raise TypeError do UniffiTypeLimits.take_u8(true) end
    assert_raise TypeError do UniffiTypeLimits.take_u16(true) end
    assert_raise TypeError do UniffiTypeLimits.take_u32(true) end
    assert_raise TypeError do UniffiTypeLimits.take_u64(true) end

    assert_raise TypeError do UniffiTypeLimits.take_i8(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_i16(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_i32(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_i64(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_u8(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_u16(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_u32(NonInteger.new) end
    assert_raise TypeError do UniffiTypeLimits.take_u64(NonInteger.new) end
  end
  class IntegerLike
    def to_int
      123
    end
  end
  def test_integer_like
    assert_equal(UniffiTypeLimits.take_i8(123.0), 123)
    assert_equal(UniffiTypeLimits.take_i16(123.0), 123)
    assert_equal(UniffiTypeLimits.take_i32(123.0), 123)
    assert_equal(UniffiTypeLimits.take_i64(123.0), 123)
    assert_equal(UniffiTypeLimits.take_u8(123.0), 123)
    assert_equal(UniffiTypeLimits.take_u16(123.0), 123)
    assert_equal(UniffiTypeLimits.take_u32(123.0), 123)
    assert_equal(UniffiTypeLimits.take_u64(123.0), 123)

    assert_equal(UniffiTypeLimits.take_i8(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_i16(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_i32(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_i64(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_u8(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_u16(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_u32(-0.5), 0)
    assert_equal(UniffiTypeLimits.take_u64(-0.5), 0)

    assert_equal(UniffiTypeLimits.take_i8(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_i16(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_i32(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_i64(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_u8(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_u16(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_u32(IntegerLike.new), 123)
    assert_equal(UniffiTypeLimits.take_u64(IntegerLike.new), 123)
  end
  class NonFloat
  end
  def test_non_float
    assert_raise TypeError do UniffiTypeLimits.take_f32(nil) end
    assert_raise TypeError do UniffiTypeLimits.take_f64(nil) end

    assert_raise TypeError do UniffiTypeLimits.take_f32("0") end
    assert_raise TypeError do UniffiTypeLimits.take_f64("0") end

    assert_raise TypeError do UniffiTypeLimits.take_f32(false) end
    assert_raise TypeError do UniffiTypeLimits.take_f64(false) end

    assert_raise TypeError do UniffiTypeLimits.take_f32(true) end
    assert_raise TypeError do UniffiTypeLimits.take_f64(true) end

    assert_raise RangeError do UniffiTypeLimits.take_f32(1i) end
    assert_raise RangeError do UniffiTypeLimits.take_f64(1i) end

    assert_raise TypeError do UniffiTypeLimits.take_f32(NonFloat.new) end
    assert_raise TypeError do UniffiTypeLimits.take_f64(NonFloat.new) end
  end
  def test_float_like
    assert_equal(UniffiTypeLimits.take_f32(456), 456.0)
    assert_equal(UniffiTypeLimits.take_f64(456), 456.0)
  end
  def test_special_floats
    assert_equal(UniffiTypeLimits.take_f32(Float::INFINITY), Float::INFINITY)
    assert_equal(UniffiTypeLimits.take_f64(Float::INFINITY), Float::INFINITY)

    assert_equal(UniffiTypeLimits.take_f32(-Float::INFINITY), -Float::INFINITY)
    assert_equal(UniffiTypeLimits.take_f64(-Float::INFINITY), -Float::INFINITY)

    assert_equal(UniffiTypeLimits.take_f32(0.0).to_s, "0.0")
    assert_equal(UniffiTypeLimits.take_f64(0.0).to_s, "0.0")

    assert_equal(UniffiTypeLimits.take_f32(-0.0).to_s, "-0.0")
    assert_equal(UniffiTypeLimits.take_f64(-0.0).to_s, "-0.0")

    assert(UniffiTypeLimits.take_f32(Float::NAN).nan?)
    assert(UniffiTypeLimits.take_f64(Float::NAN).nan?)
  end
  def test_strings
    assert_raise Encoding::InvalidByteSequenceError do UniffiTypeLimits.take_string("\xff") end # invalid byte
    assert_raise Encoding::InvalidByteSequenceError do UniffiTypeLimits.take_string("\xed\xa0\x80") end # surrogate
    assert_equal(UniffiTypeLimits.take_string(""), "")
    assert_equal(UniffiTypeLimits.take_string("愛"), "愛")
    assert_equal(UniffiTypeLimits.take_string("💖"), "💖")
  end
end
